File: BeeCrowd/1231-Words/solution.cpp
/**
 * Problem: 1231 - Words
 * Language: C++17 (Optimized)
 * Approach: Static Trie Arrays + DFS with 2D Visited Array
 */

#include <iostream>
#include <vector>
#include <string>
#include <cstring>

using namespace std;

// Structure to represent a Trie Node
struct Node
{
    int children[2]; // Indices of children nodes. -1 if null.
    bool isEnd;      // True if this node marks the end of a word.

    Node()
    {
        children[0] = -1;
        children[1] = -1;
        isEnd = false;
    }
};

// We use two vectors to act as our memory pools for the two Tries.
// This avoids the overhead of 'new' and pointer chasing.
vector<Node> trie1;
vector<Node> trie2;

// Visited array for Memoization.
// Max nodes calculation: 20 words * 40 chars = 800 nodes max.
// We use 1000 to be safe.
bool visited[1000][1000];

// Helper to insert words into our vector-based Trie
void insert(vector<Node> &trie, const string &s)
{
    int curr = 0; // Start at root (index 0)
    for (char c : s)
    {
        int bit = c - '0';
        if (trie[curr].children[bit] == -1)
        {
            // Create new node
            Node newNode;
            trie.push_back(newNode);
            trie[curr].children[bit] = trie.size() - 1; // Link to new index
        }
        curr = trie[curr].children[bit];
    }
    trie[curr].isEnd = true;
}

// DFS Function
// u: current index in trie1
// v: current index in trie2
bool dfs(int u, int v)
{
    // 1. Check for Cycle / Already Visited State
    if (visited[u][v])
        return false;
    visited[u][v] = true;

    // 2. Check for Successful Match
    // If both nodes mark the end of a word, we have found a valid common sequence.
    if (trie1[u].isEnd && trie2[v].isEnd)
        return true;

    // 3. Try transitions for '0' and '1'
    for (int k = 0; k < 2; k++)
    {
        int nextU = trie1[u].children[k];
        int nextV = trie2[v].children[k];

        // If both tries have a transition for this bit k
        if (nextU != -1 && nextV != -1)
        {
            if (dfs(nextU, nextV))
                return true;
        }
    }

    // 4. Epsilon Transitions (Restart Logic)
    // If trie1 ends a word here, we can conceptually jump 'u' back to root (0)
    // while keeping 'v' at the same place.
    if (trie1[u].isEnd)
    {
        if (dfs(0, v))
            return true;
    }

    // If trie2 ends a word here, jump 'v' back to root (0).
    if (trie2[v].isEnd)
    {
        if (dfs(u, 0))
            return true;
    }

    return false;
}

int main()
{
    // Fast I/O
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n1, n2;
    while (cin >> n1 >> n2)
    {
        // Clear previous test case data
        trie1.clear();
        trie2.clear();

        // Initialize roots
        trie1.push_back(Node());
        trie2.push_back(Node());

        string s;
        for (int i = 0; i < n1; i++)
        {
            cin >> s;
            insert(trie1, s);
        }
        for (int i = 0; i < n2; i++)
        {
            cin >> s;
            insert(trie2, s);
        }

        // Reset visited array
        // We only need to clear the area we might use.
        // memset is very fast.
        // Size to clear: trie1.size() rows * 1000 cols.
        for (int i = 0; i < trie1.size(); ++i)
        {
            memset(visited[i], 0, trie2.size() * sizeof(bool));
        }

        if (dfs(0, 0))
        {
            cout << "S" << "\n";
        }
        else
        {
            cout << "N" << "\n";
        }
    }

    return 0;
}

File: BeeCrowd/1231-Words/solution.js
"use strict";

// Use a large buffer to read input quickly
const input = require("fs").readFileSync("/dev/stdin", "utf8");
const lines = input.trim().split(/\s+/); // Split by any whitespace

/**
 * Problem: 1231 - Words
 * Language: JavaScript (Node.js)
 * Approach: Simultaneous DFS on Two Tries with Memoization
 */

// Global counter to assign unique IDs to every node for fast state caching
let nodeIdCounter = 0;

class Node {
  constructor() {
    // Optimization: Use Array(2) instead of Object for children.
    // Index 0 for '0', Index 1 for '1'.
    this.children = [null, null];
    this.isEnd = false;
    this.id = nodeIdCounter++; // Unique integer ID for this node
  }
}

class Trie {
  constructor() {
    this.root = new Node();
  }

  insert(word) {
    let node = this.root;
    for (let i = 0; i < word.length; i++) {
      // Convert char '0'/'1' to integer 0/1
      const idx = word.charCodeAt(i) - 48;
      if (!node.children[idx]) {
        node.children[idx] = new Node();
      }
      node = node.children[idx];
    }
    node.isEnd = true;
  }
}

function solve() {
  let cursor = 0;

  // Process all test cases
  while (cursor < lines.length) {
    const N1 = parseInt(lines[cursor++]);
    const N2 = parseInt(lines[cursor++]);

    // Reset global ID counter for each test case to keep IDs small
    nodeIdCounter = 0;

    const trie1 = new Trie();
    const trie2 = new Trie();

    for (let i = 0; i < N1; i++) trie1.insert(lines[cursor++]);
    for (let i = 0; i < N2; i++) trie2.insert(lines[cursor++]);

    // Visited Set to store states we've already processed.
    // State key: a combination of node1.id and node2.id
    const visited = new Set();

    function dfs(n1, n2) {
      // 1. Successful Match Condition:
      // If both pointers are at the end of a word simultaneously, we found a match.
      // Note: We can return true immediately, OR we could continue to see if
      // longer matches exist. For this problem, finding *one* is enough.
      if (n1.isEnd && n2.isEnd) return true;

      // 2. Memoization / Cycle Detection:
      // Generate a unique key for the state (n1, n2).
      // Since we have max ~800 nodes (20 words * 40 chars), IDs fit in 11 bits.
      // (n1.id << 12) | n2.id creates a unique integer hash.
      const stateKey = (n1.id << 12) | n2.id;

      if (visited.has(stateKey)) return false;
      visited.add(stateKey);

      // 3. Transitions: Try to move both pointers forward with '0' and '1'

      // Try matching '0'
      if (n1.children[0] && n2.children[0]) {
        if (dfs(n1.children[0], n2.children[0])) return true;
      }

      // Try matching '1'
      if (n1.children[1] && n2.children[1]) {
        if (dfs(n1.children[1], n2.children[1])) return true;
      }

      // 4. Epsilon Transitions (Word Boundaries):
      // If we reached the end of a word in Set 1, we can "restart" traversal
      // at the root of Set 1, effectively concatenating a new word from Set 1.
      // We keep n2 at its current position.
      if (n1.isEnd) {
        if (dfs(trie1.root, n2)) return true;
      }

      // Same logic for Set 2
      if (n2.isEnd) {
        if (dfs(n1, trie2.root)) return true;
      }

      return false;
    }

    // Start DFS from the roots of both Tries
    if (dfs(trie1.root, trie2.root)) {
      process.stdout.write("S\n");
    } else {
      process.stdout.write("N\n");
    }
  }
}

solve();
